File: cars_pack/src/ModelBartlet.cpp
#include "ModelBartlet.h"

#include <algorithm>
#include <iostream>
#include <cmath>

#include "ModelPoisson.h"

ModelBartlet::ModelBartlet() : r(0), g(0) {
}

ModelBartlet::ModelBartlet(double r_, double g_) : r(r_), g(g_) {
}

size_t ModelBartlet::countRequests() {
    double p = generateProbability();
    std::cout << "Random: " << p;
    int count = 0;
    double F = 1.0 - r;
    std::cout << " F: " << F;
    double l = 0.0;
    double temp_g = 1;
    double temp = r * (1.0 - g);
    while (F <= p) {
        count++;
        l = temp * temp_g;
        F += l;
        if (std::abs(l) <= eps)
            return count + 1ull;
        temp_g *= g;
    }
    std::cout << " Count: " << count << std::endl;
    return count;
}

double ModelBartlet::getExpectedValue() const {
    return 1 + (r / 1 - g);
}

double ModelBartlet::getExpectedValue(double r_, double g_) const {
    return 1 + (r_ / 1 - g_);
}

double ModelBartlet::getProbably(size_t k) {
    if (k == 1)
        return 1 - r;
    else if (k >= 2)
        return r * (1 - g) * pow(g, k - 2);
    return 0;
}

File: cars_pack_py/backend/model_bartlet.py
import random

from .model import Model
from . import consts


class ModelBartlet(Model):
    def __init__(self, r=0, g=0) -> None:
        super().__init__()
        if not (0 <= r <= 1 and 0 <= g <= 1):
            raise ValueError()
        self.r = r
        self.g = g

    @staticmethod
    def get_expected_value_custom(r: float, g: float) -> float:
        return 1 + (r / (1 - g))

    def get_expected_value(self) -> float:
        return 1 + (self.r / (1 - self.g))

    def get_probability(self, k: int) -> float:
        if k == 1:
            return 1 - self.r
        if k >= 2:
            return self.r * (1 - self.g) * self.g ** (k - 2)
        return 0

    def count_requests(self) -> int:
        count = 0
        p = random.uniform(0, 1 - consts.EPSILON)
        temp = self.r * (1 - self.g)
        F = 1 - self.r
        while F <= p:
            count += 1
            F += temp
            temp *= self.g
            if abs(temp) <= consts.EPSILON:
                return count + 1

        return count
